crates/iceberg/src/transaction/rewrite_files.rs

```diff
@@ -0,0 +1,375 @@
+                    entry.file_path()
+                ),
+            )
+        })?)
+        // todo copy file seq no as well
+        .data_file(entry.data_file().clone());
+
+    Ok(builder.build())
+}
+
+impl SnapshotValidator for RewriteFilesOperation {
+    async fn validate(&self, base: &Table, parent_snapshot_id: Option<i64>) -> Result<()> {
+        // Validate replaced and added files
+        if self.deleted_data_files.is_empty() && self.deleted_delete_files.is_empty() {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                "Files to delete cannot be empty",
+            ));
+        }
+        if self.deleted_data_files.is_empty() && !self.added_data_files.is_empty() {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                "Data files to add must be empty because there's no data file to be rewritten",
+            ));
+        }
+        if self.deleted_delete_files.is_empty() && !self.added_delete_files.is_empty() {
+            return Err(Error::new(
+                ErrorKind::DataInvalid,
+                "Delete files to add must be empty because there's no delete file to be rewritten",
+            ));
+        }
+
+        // todo add use_starting_seq_number to determine if we want to use data_sequence_number
+        // If there are replaced data files, there cannot be any new row-level deletes for those data files
+        if !self.deleted_data_files.is_empty() {
+            self.validate_no_new_deletes_for_data_files(
+                base,
+                self.starting_snapshot_id,
+                parent_snapshot_id,
+                &self.deleted_data_files,
+                self.data_sequence_number.is_some(),
+            )
+            .await?;
+        }
+
+        Ok(())
+    }
+}
+
+impl SnapshotProduceOperation for RewriteFilesOperation {
+    fn operation(&self) -> Operation {
+        Operation::Replace
+    }
+
+    async fn delete_entries(
+        &self,
+        snapshot_producer: &SnapshotProducer<'_>,
+    ) -> Result<Vec<ManifestEntry>> {
+        // Find entries that are associated with deleted files
+        let snapshot = snapshot_producer.table.metadata().current_snapshot();
+
+        if let Some(snapshot) = snapshot {
+            let manifest_list = snapshot
+                .load_manifest_list(
+                    snapshot_producer.table.file_io(),
+                    snapshot_producer.table.metadata(),
+                )
+                .await?;
+
+            let mut delete_entries = Vec::new();
+
+            for manifest_file in manifest_list.entries() {
+                let manifest = manifest_file
+                    .load_manifest(snapshot_producer.table.file_io())
+                    .await?;
+
+                for entry in manifest.entries() {
+                    match entry.content_type() {
+                        DataContentType::Data => {
```

Comment:
These are the exact same code blocks? So why does the `ManifestEntry`'s `DataContentType` matter? Why not just do:
```rust
for entry in manifest.entries() {
    if snapshot_producer
        .deleted_data_files
        .iter()
        .any(|f| f.file_path == entry.data_file.file_path())
     {
         delete_entries.push(copy_with_deleted_status())
     }
        
}
```